web/src/app/services/diagram/waypoint-manager.service.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  distinctUntilChanged,
+  map,
+  Observable,
+  ReplaySubject,
+  Subject,
+} from 'rxjs';
+import { DiagramWaypointAreaDirective } from './waypoint-area.directive';
+
+/**
+ * WaypointArea represents rectangular area on the diagram where can be used as a connector of lines.
+ */
+
+export interface WaypointAreaViewportSpace {
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+}
+
+/**
+ * WaypointManagerService manages the points on diagrams.
+ * Arrows or lines uses the named points to draw.
+ */
+export class WaypointManagerService {
+  private readonly waypointAreaSubjects: { [id: string]: Subject<DOMRect> } =
+    {};
+
+  /**
+   * monitorWaypoint get the observable to monitor a point at the specified area.
+   * positionInArea is [0,1] normalized DOMPoint locates the relative location in the area. [0,0] means top left, [0,1] means bottom left for example.
+   */
```

Comment:
If [0,0] means top left, I assume [top, bottom] = [0,1] and [left,right] = [0, 1].
Then [0,1] should be top right instead of bottom left mentioned here? I might miss something

Reply:
The array is in the form `[x,y]`, where `x` and `y` are values normalized relative to their parent's dimensions.

* `[0,y]` means left and `[1,y`] means right
* `[x,0]` means top and `[x,1] means bottom
Thus `[0,0]` means left top and `[0,1]` means bottom left

Reply:
I see, thanks for expalnation!